clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsGroupHeartbeatRequestManager.java

```diff
@@ -0,0 +1,387 @@
+            final StreamsRebalanceData.Subtopology subtopologyDataFromStreams) {
+            final List<StreamsGroupHeartbeatRequestData.TopicInfo> changelogTopicsInfo = new ArrayList<>();
+            for (final Map.Entry<String, StreamsRebalanceData.TopicInfo> changelogTopic : subtopologyDataFromStreams.stateChangelogTopics().entrySet()) {
+                final StreamsGroupHeartbeatRequestData.TopicInfo changelogTopicInfo = new StreamsGroupHeartbeatRequestData.TopicInfo();
+                changelogTopicInfo.setName(changelogTopic.getKey());
+                changelogTopic.getValue().replicationFactor().ifPresent(changelogTopicInfo::setReplicationFactor);
+                changelogTopic.getValue().topicConfigs().forEach((k, v) ->
+                    changelogTopicInfo.topicConfigs().add(new StreamsGroupHeartbeatRequestData.KeyValue().setKey(k).setValue(v))
+                );
+                changelogTopicsInfo.add(changelogTopicInfo);
+            }
+            changelogTopicsInfo.sort(Comparator.comparing(StreamsGroupHeartbeatRequestData.TopicInfo::name));
+            return changelogTopicsInfo;
+        }
+    }
+
+
+    /**
+     * Represents the state of a heartbeat request, including logic for timing, retries, and exponential backoff. The object extends
+     * {@link RequestState} to enable exponential backoff and duplicated request handling. The two fields that it holds are:
+     */
+    static class HeartbeatRequestState extends RequestState {
+
+        /**
+         * heartbeatTimer tracks the time since the last heartbeat was sent
+         */
+        private final Timer heartbeatTimer;
+
+        /**
+         * The heartbeat interval which is acquired/updated through the heartbeat request
+         */
+        private long heartbeatIntervalMs;
+
+        public HeartbeatRequestState(
+            final LogContext logContext,
+            final Time time,
+            final long heartbeatIntervalMs,
+            final long retryBackoffMs,
+            final long retryBackoffMaxMs,
+            final double jitter) {
+            super(logContext, StreamsGroupHeartbeatRequestManager.HeartbeatRequestState.class.getName(), retryBackoffMs, 2,
+                retryBackoffMaxMs,
+                jitter);
+            this.heartbeatIntervalMs = heartbeatIntervalMs;
+            this.heartbeatTimer = time.timer(heartbeatIntervalMs);
+        }
+
+        private void update(final long currentTimeMs) {
+            this.heartbeatTimer.update(currentTimeMs);
+        }
+
+        public void resetTimer() {
+            this.heartbeatTimer.reset(heartbeatIntervalMs);
+        }
+
+        @Override
+        public boolean canSendRequest(final long currentTimeMs) {
+            update(currentTimeMs);
+            return heartbeatTimer.isExpired() && super.canSendRequest(currentTimeMs);
+        }
+
+        private void updateHeartbeatIntervalMs(final long heartbeatIntervalMs) {
```

Comment:
Same here.

Reply:
I was able to verify this by polling again the heartbeat request manager at the end of the test and checking the new time until the next poll.

However, verification for the other methods you commented on it hard since they are on internal objects. I will try to refactor `HeartbeatRequestState` to an external class in a separate PR, since it is duplicated in the `ConsumerHeartbeatRequestManager`. In the course of that refactoring, I will also test the code you mentioned.